In [2]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)

df = pd.read_csv("../data/raw/fraud_oracle.csv")   # replace with your real filename
print(df.shape)
df.head()

FileNotFoundError: [Errno 2] No such file or directory: '../data/raw/fraud_oracle.csv'

In [3]:
import os
print(os.listdir("../data/raw"))

FileNotFoundError: [WinError 3] The system cannot find the path specified: '../data/raw'

In [4]:
import os
print(os.getcwd())

C:\Users\N6N0C\Documents\insurance-fraud-detection


In [1]:
import os
print(os.listdir("../data/raw"))

['fraud_oracle.csv']


In [2]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)

df = pd.read_csv("../data/raw/fraud_oracle.csv")   # adjust if the exact name differs
print(df.shape)
df.head()

(15420, 33)


,Month,WeekOfMonth,DayOfWeek,Make,AccidentArea,DayOfWeekClaimed,MonthClaimed,WeekOfMonthClaimed,Sex,MaritalStatus,Age,Fault,PolicyType,VehicleCategory,VehiclePrice,FraudFound_P,PolicyNumber,RepNumber,Deductible,DriverRating,Days_Policy_Accident,Days_Policy_Claim,PastNumberOfClaims,AgeOfVehicle,AgeOfPolicyHolder,PoliceReportFiled,WitnessPresent,AgentType,NumberOfSuppliments,AddressChange_Claim,NumberOfCars,Year,BasePolicy
0,Dec,5,Wednesday,Honda,Urban,Tuesday,Jan,1,Female,Single,21,Policy Holder,Sport - Liability,Sport,more than 69000,0,1,12,300,1,more than 30,more than 30,none,3 years,26 to 30,No,No,External,none,1 year,3 to 4,1994,Liability
1,Jan,3,Wednesday,Honda,Urban,Monday,Jan,4,Male,Single,34,Policy Holder,Sport - Collision,Sport,more than 69000,0,2,15,400,4,more than 30,more than 30,none,6 years,31 to 35,Yes,No,External,none,no change,1 vehicle,1994,Collision
2,Oct,5,Friday,Honda,Urban,Thursday,Nov,2,Male,Married,47,Policy Holder,Sport - Collision,Sport,more than 69000,0,3,7,400,3,more than 30,more than 30,1,7 years,41 to 50,No,No,External,none,no change,1 vehicle,1994,Collision
3,Jun,2,Saturday,Toyota,Rural,Friday,Jul,1,Male,Married,65,Third Party,Sedan - Liability,Sport,20000 to 29000,0,4,4,400,2,more than 30,more than 30,1,more than 7,51 to 65,Yes,No,External,more than 5,no change,1 vehicle,1994,Liability
4,Jan,5,Monday,Honda,Urban,Tuesday,Feb,2,Female,Single,27,Third Party,Sport - Collision,Sport,more than 69000,0,5,3,400,1,more than 30,more than 30,none,5 years,31 to 35,No,No,External,none,no change,1 vehicle,1994,Collision


In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 15420 entries, 0 to 15419
Data columns (total 33 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   Month                 15420 non-null  str  
 1   WeekOfMonth           15420 non-null  int64
 2   DayOfWeek             15420 non-null  str  
 3   Make                  15420 non-null  str  
 4   AccidentArea          15420 non-null  str  
 5   DayOfWeekClaimed      15420 non-null  str  
 6   MonthClaimed          15420 non-null  str  
 7   WeekOfMonthClaimed    15420 non-null  int64
 8   Sex                   15420 non-null  str  
 9   MaritalStatus         15420 non-null  str  
 10  Age                   15420 non-null  int64
 11  Fault                 15420 non-null  str  
 12  PolicyType            15420 non-null  str  
 13  VehicleCategory       15420 non-null  str  
 14  VehiclePrice          15420 non-null  str  
 15  FraudFound_P          15420 non-null  int64
 16  PolicyNumber   

In [4]:
df.isna().sum().sort_values(ascending=False)

Month                   0
WeekOfMonth             0
DayOfWeek               0
Make                    0
AccidentArea            0
DayOfWeekClaimed        0
MonthClaimed            0
WeekOfMonthClaimed      0
Sex                     0
MaritalStatus           0
Age                     0
Fault                   0
PolicyType              0
VehicleCategory         0
VehiclePrice            0
FraudFound_P            0
PolicyNumber            0
RepNumber               0
Deductible              0
DriverRating            0
Days_Policy_Accident    0
Days_Policy_Claim       0
PastNumberOfClaims      0
AgeOfVehicle            0
AgeOfPolicyHolder       0
PoliceReportFiled       0
WitnessPresent          0
AgentType               0
NumberOfSuppliments     0
AddressChange_Claim     0
NumberOfCars            0
Year                    0
BasePolicy              0
dtype: int64

In [5]:
print(df["FraudFound_P"].value_counts())
print(df["FraudFound_P"].value_counts(normalize=True).round(3))

FraudFound_P
0    14497
1      923
Name: count, dtype: int64
FraudFound_P
0    0.94
1    0.06
Name: proportion, dtype: float64


In [6]:
print("Duplicate rows:", df.duplicated().sum())
print("Columns:", df.columns.tolist())

Duplicate rows: 0
Columns: ['Month', 'WeekOfMonth', 'DayOfWeek', 'Make', 'AccidentArea', 'DayOfWeekClaimed', 'MonthClaimed', 'WeekOfMonthClaimed', 'Sex', 'MaritalStatus', 'Age', 'Fault', 'PolicyType', 'VehicleCategory', 'VehiclePrice', 'FraudFound_P', 'PolicyNumber', 'RepNumber', 'Deductible', 'DriverRating', 'Days_Policy_Accident', 'Days_Policy_Claim', 'PastNumberOfClaims', 'AgeOfVehicle', 'AgeOfPolicyHolder', 'PoliceReportFiled', 'WitnessPresent', 'AgentType', 'NumberOfSuppliments', 'AddressChange_Claim', 'NumberOfCars', 'Year', 'BasePolicy']


In [7]:
print(df["PolicyNumber"].nunique(), "unique values out of", len(df), "rows")
print(df["RepNumber"].value_counts().head(10))
print(df["Year"].value_counts())
print(df["Days_Policy_Accident"].value_counts())
print(df["PastNumberOfClaims"].value_counts())
print(df["AgeOfVehicle"].value_counts())
print(df["PoliceReportFiled"].value_counts())
print(df["WitnessPresent"].value_counts())

15420 unique values out of 15420 rows
RepNumber
7     1069
9      999
1      987
5      987
10     986
12     977
15     977
16     967
2      956
3      949
Name: count, dtype: int64
Year
1994    6142
1995    5195
1996    4083
Name: count, dtype: int64
Days_Policy_Accident
more than 30    15247
none               55
8 to 15            55
15 to 30           49
1 to 7             14
Name: count, dtype: int64
PastNumberOfClaims
2 to 4         5485
none           4352
1              3573
more than 4    2010
Name: count, dtype: int64
AgeOfVehicle
7 years        5807
more than 7    3981
6 years        3448
5 years        1357
new             373
4 years         229
3 years         152
2 years          73
Name: count, dtype: int64
PoliceReportFiled
No     14992
Yes      428
Name: count, dtype: int64
WitnessPresent
No     15333
Yes       87
Name: count, dtype: int64


## Data Overview
- 15,420 rows, 33 columns, 0 missing values, 0 duplicates — a clean dataset
- Target: FraudFound_P (0 = no fraud, 1 = fraud)
- Fraud rate: 6.0% (923 of 15,420 claims) — imbalanced; precision/recall/PR-AUC prioritized over accuracy

### Column notes
- PolicyNumber: 15,420/15,420 unique — pure ID, excluded from modeling
- RepNumber: 16 categories, evenly distributed — kept as categorical, weak predictor expected
- Year: only 1994-1996 — old dataset; noted as a limitation
- Days_Policy_Accident: 98.9% "more than 30" — very low variance, likely weak predictor
- PastNumberOfClaims, AgeOfVehicle: ordinal categories, need custom sort order for charts
- PoliceReportFiled: 2.8% "Yes" — tests H4, but rare-category imbalance itself
- WitnessPresent: 0.6% "Yes" — extremely rare, limited statistical power expected

### Known Limitations
- Data is from 1994-1996 — a synthetic/older dataset, not representative of current fraud patterns
- FraudFound_P imbalance (6%) requires careful metric choice (precision, recall, PR-AUC, not accuracy)
- Findings show association, not proof that a factor causes fraud

In [8]:
df_clean = df.copy()
print("Start:", df_clean.shape)

Start: (15420, 33)


In [10]:
cat_cols = df_clean.select_dtypes(include=["object", "str"]).columns.tolist()
for col in cat_cols:
    print(f"\n--- {col} ({df_clean[col].nunique()} unique) ---")
    print(df_clean[col].value_counts())


--- Month (12 unique) ---
Month
Jan    1411
May    1367
Mar    1360
Jun    1321
Oct    1305
Dec    1285
Apr    1280
Feb    1266
Jul    1257
Sep    1240
Nov    1201
Aug    1127
Name: count, dtype: int64

--- DayOfWeek (7 unique) ---
DayOfWeek
Monday       2616
Friday       2445
Tuesday      2300
Thursday     2173
Wednesday    2159
Saturday     1982
Sunday       1745
Name: count, dtype: int64

--- Make (19 unique) ---
Make
Pontiac      3837
Toyota       3121
Honda        2801
Mazda        2354
Chevrolet    1681
Accura        472
Ford          450
VW            283
Dodge         109
Saab          108
Mercury        83
Saturn         58
Nisson         30
BMW            15
Jaguar          6
Porche          5
Mecedes         4
Ferrari         2
Lexus           1
Name: count, dtype: int64

--- AccidentArea (2 unique) ---
AccidentArea
Urban    13822
Rural     1598
Name: count, dtype: int64

--- DayOfWeekClaimed (8 unique) ---
DayOfWeekClaimed
Monday       3757
Tuesday      3375
Wednesday    2

In [11]:
num_cols = df_clean.select_dtypes(include=["int64", "float64"]).columns.tolist()
df_clean[num_cols].describe()

,WeekOfMonth,WeekOfMonthClaimed,Age,FraudFound_P,PolicyNumber,RepNumber,Deductible,DriverRating,Year
count,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000
mean,2.788586,2.693969,39.855707,0.059857,7710.500000,8.483268,407.704280,2.487808,1994.866472
std,1.287585,1.259115,13.492377,0.237230,4451.514911,4.599948,43.950998,1.119453,0.803313
min,1.000000,1.000000,0.000000,0.000000,1.000000,1.000000,300.000000,1.000000,1994.000000
25%,2.000000,2.000000,31.000000,0.000000,3855.750000,5.000000,400.000000,1.000000,1994.000000
50%,3.000000,3.000000,38.000000,0.000000,7710.500000,8.000000,400.000000,2.000000,1995.000000
75%,4.000000,4.000000,48.000000,0.000000,11565.250000,12.000000,400.000000,3.000000,1996.000000
max,5.000000,5.000000,80.000000,1.000000,15420.000000,16.000000,700.000000,4.000000,1996.000000


In [12]:
print("PolicyNumber uniqueness confirmed:", df_clean["PolicyNumber"].nunique() == len(df_clean))
# No action needed here — we exclude it later at modeling time, not now

PolicyNumber uniqueness confirmed: True


In [13]:
age_vehicle_order = ["new", "2 years", "3 years", "4 years", "5 years", "6 years", "7 years", "more than 7"]
past_claims_order = ["none", "1", "2 to 4", "more than 4"]
days_policy_order = ["none", "1 to 7", "8 to 15", "15 to 30", "more than 30"]

df_clean["AgeOfVehicle"] = pd.Categorical(df_clean["AgeOfVehicle"], categories=age_vehicle_order, ordered=True)
df_clean["PastNumberOfClaims"] = pd.Categorical(df_clean["PastNumberOfClaims"], categories=past_claims_order, ordered=True)
df_clean["Days_Policy_Accident"] = pd.Categorical(df_clean["Days_Policy_Accident"], categories=days_policy_order, ordered=True)

# Verify no values got dropped/mismatched (would show as NaN if a label didn't match exactly)
print(df_clean["AgeOfVehicle"].isna().sum())
print(df_clean["PastNumberOfClaims"].isna().sum())
print(df_clean["Days_Policy_Accident"].isna().sum())

0
0
0


In [14]:
df_clean.to_csv("../data/cleaned/fraud_clean.csv", index=False)
print("Saved:", df_clean.shape)

Saved: (15420, 33)


## Cleaning Log
- Raw: 15,420 rows, 33 columns, 0 missing, 0 duplicates
- Scanned all 24 categorical columns for typos/inconsistent values — [note what you found]
- Confirmed PolicyNumber is a pure ID (15,420/15,420 unique) — excluded from modeling features only
- Set ordinal category order for AgeOfVehicle, PastNumberOfClaims, Days_Policy_Accident
- No rows dropped in this step — data was already clean
- Saved to data/cleaned/fraud_clean.csv

In [15]:
print("Age == 0:", (df_clean["Age"] == 0).sum())
print(df_clean[["DayOfWeekClaimed", "MonthClaimed"]].apply(lambda s: (s == "0").sum()))

Age == 0: 320
DayOfWeekClaimed    1
MonthClaimed        1
dtype: int64


In [16]:
# Age = 0 is a placeholder for "unknown", not a real age
print(pd.crosstab(df_clean["Age"] == 0, df_clean["AgeOfPolicyHolder"]))

df_clean["Age_Missing"] = (df_clean["Age"] == 0).astype(int)
df_clean["Age_Clean"] = df_clean["Age"].replace(0, np.nan)

# Drop rows with an invalid claim month/day (very few, if any)
before = len(df_clean)
df_clean = df_clean[(df_clean["MonthClaimed"] != "0") & (df_clean["DayOfWeekClaimed"] != "0")]
print("Rows dropped:", before - len(df_clean))

AgeOfPolicyHolder  16 to 17  18 to 20  21 to 25  26 to 30  31 to 35  36 to 40  \
Age                                                                             
False                     0        15       108       613      5593      4043   
True                    320         0         0         0         0         0   

AgeOfPolicyHolder  41 to 50  51 to 65  over 65  
Age                                             
False                  2828      1392      508  
True                      0         0        0  
Rows dropped: 1


In [17]:
ordinal_orders = {
    "AgeOfVehicle": ["new", "2 years", "3 years", "4 years", "5 years", "6 years", "7 years", "more than 7"],
    "PastNumberOfClaims": ["none", "1", "2 to 4", "more than 4"],
    "Days_Policy_Accident": ["none", "1 to 7", "8 to 15", "15 to 30", "more than 30"],
    "VehiclePrice": ["less than 20000", "20000 to 29000", "30000 to 39000", "40000 to 59000", "60000 to 69000", "more than 69000"],
    "AgeOfPolicyHolder": ["16 to 17", "18 to 25", "26 to 30", "31 to 35", "36 to 40", "41 to 50", "51 to 65", "over 65"],
    "NumberOfSuppliments": ["none", "1 to 2", "3 to 5", "more than 5"],
    "AddressChange_Claim": ["no change", "under 6 months", "1 year", "2 to 3 years", "4 to 8 years"],
    "Days_Policy_Claim": ["none", "8 to 15", "15 to 30", "more than 30"],
}

for col, order in ordinal_orders.items():
    df_clean[col + "_ord"] = pd.Categorical(df_clean[col], categories=order, ordered=True).codes
    bad = (df_clean[col + "_ord"] == -1).sum()
    print(f"{col}: unmatched = {bad}")
    if bad > 0:
        print("   actual labels:", df_clean[col].unique().tolist())

AgeOfVehicle: unmatched = 0
PastNumberOfClaims: unmatched = 0
Days_Policy_Accident: unmatched = 0
VehiclePrice: unmatched = 0
AgeOfPolicyHolder: unmatched = 123
   actual labels: ['26 to 30', '31 to 35', '41 to 50', '51 to 65', '21 to 25', '36 to 40', '16 to 17', 'over 65', '18 to 20']
NumberOfSuppliments: unmatched = 0
AddressChange_Claim: unmatched = 0
Days_Policy_Claim: unmatched = 0


C:\Users\N6N0C\AppData\Local\Temp\ipykernel_14216\3533877389.py:13: Pandas4Warning: Constructing a Categorical with a dtype and values containing non-null entries not in that dtype's categories is deprecated and will raise in a future version.
  df_clean[col + "_ord"] = pd.Categorical(df_clean[col], categories=order, ordered=True).codes


In [18]:
months = ["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"]
m = {name: i + 1 for i, name in enumerate(months)}

df_clean["Claim_Delay_Months"] = (df_clean["MonthClaimed"].map(m) - df_clean["Month"].map(m)) % 12
df_clean["Weekend_Accident"] = df_clean["DayOfWeek"].isin(["Saturday", "Sunday"]).astype(int)
df_clean["Weekend_Claim"] = df_clean["DayOfWeekClaimed"].isin(["Saturday", "Sunday"]).astype(int)

print(df_clean["Claim_Delay_Months"].value_counts().sort_index())

Claim_Delay_Months
0     11442
1      3211
2       390
3       145
4        80
5        38
6        37
7        29
8        23
9        10
10        7
11        7
Name: count, dtype: int64


In [19]:
df_clean["Police_Report"] = (df_clean["PoliceReportFiled"] == "Yes").astype(int)
df_clean["Witness"] = (df_clean["WitnessPresent"] == "Yes").astype(int)
df_clean["Policyholder_At_Fault"] = (df_clean["Fault"] == "Policy Holder").astype(int)
df_clean["Agent_External"] = (df_clean["AgentType"] == "External").astype(int)

print(df_clean[["Police_Report", "Witness", "Policyholder_At_Fault", "Agent_External"]].mean().round(3))

Police_Report            0.028
Witness                  0.006
Policyholder_At_Fault    0.728
Agent_External           0.984
dtype: float64


In [20]:
for col in ["PoliceReportFiled", "WitnessPresent", "Fault", "PolicyType", "BasePolicy", "AccidentArea", "AgentType"]:
    print(df_clean.groupby(col)["FraudFound_P"].agg(["count", "mean"]).round(3))
    print()

                   count   mean
PoliceReportFiled              
No                 14991  0.061
Yes                  428  0.037

                count   mean
WitnessPresent              
No              15332  0.060
Yes                87  0.034

               count   mean
Fault                      
Policy Holder  11229  0.079
Third Party     4190  0.009

                      count   mean
PolicyType                        
Sedan - All Perils     4086  0.101
Sedan - Collision      5584  0.069
Sedan - Liability      4987  0.007
Sport - All Perils       22  0.000
Sport - Collision       348  0.138
Sport - Liability         1  0.000
Utility - All Perils    340  0.121
Utility - Collision      30  0.100
Utility - Liability      21  0.000

            count   mean
BasePolicy              
All Perils   4448  0.102
Collision    5962  0.073
Liability    5009  0.007

              count   mean
AccidentArea              
Rural          1597  0.083
Urban         13822  0.057

           count   m

Every feature must be known when the claim is filed, before any investigation. If you used something learned during the investigation, the model would cheat. Everything above passes: demographics, policy details, accident details and claim timing are all known at filing. Excluded from modeling: PolicyNumber (an ID) and FraudFound_P (the target). Treat RepNumber and Year with caution: Year covers only 1994-1996, so a model could learn year-specific quirks


In [21]:
df_clean.to_csv("../data/cleaned/fraud_features.csv", index=False)
print("Saved:", df_clean.shape)

Saved: (15419, 50)


## Feature Dictionary

| Feature | Definition | Why it's needed |
|---|---|---|
| `Age_Clean` / `Age_Missing` | Age with impossible 0s set to missing / flag for those rows | Keeps unknown ages from being read as "age 0" |
| `*_ord` columns | Numeric rank of ordered categories (vehicle age, claim history, price, etc.) | Lets models and correlation tests use the natural order |
| `Claim_Delay_Months` | Months between accident and claim filing | Tests whether late claims are riskier |
| `Weekend_Accident`, `Weekend_Claim` | 1 if on Saturday/Sunday | Timing patterns |
| `Police_Report`, `Witness`, `Policyholder_At_Fault`, `Agent_External` | Yes/no flags | Tests H4 and other hypotheses |

**Leakage check:** all features are available at claim filing time. Excluded: PolicyNumber (ID), FraudFound_P (target).